File: app/src/main/java/com/alderferstudios/noteappexample/MainActivity.java
package com.alderferstudios.noteappexample;

import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;
import android.view.WindowManager;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ListView;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    //the list of items
    private List<String> listItems;
    private ArrayAdapter<String> adapter;

    private SharedPreferences prefs;
    private SharedPreferences.Editor editor;

    /**
     * Executes on app start
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));

        prefs = getPreferences(Context.MODE_PRIVATE);
        editor = prefs.edit();

        //read in saved list
        String savedList = prefs.getString("list", "");
        if (savedList.equals("")) { //nothing saved
            listItems = new ArrayList<>();
        } else {    //restore saved
            Type type = new TypeToken<ArrayList<String>>(){}.getType();
            listItems = new Gson().fromJson(savedList, type);
        }

        ListView list = (ListView) findViewById(R.id.list);

        // adapter to tie listItems to list
        adapter = new ArrayAdapter<>(this,
                R.layout.list_row, R.id.listText, listItems);

        // assign the list adapter
        list.setAdapter(adapter);
    }

    /**
     * Save list on app close
     */
    @Override
    public void onStop() {
        super.onStop();

        //turn list into json
        String items = new Gson().toJson(listItems);

        //save list to cache
        editor.putString("list", items);
        editor.apply(); //use apply, its faster than commit
    }

    /**
     * Adds an item to the ListView
     * @param v the FAB
     */
    public void addItem(View v) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Please enter the note to add");

        final EditText input = new EditText(this);
        builder.setView(input);

        builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int whichButton) {
                //add the input to the list
                listItems.add(input.getText().toString());
                adapter.notifyDataSetChanged();
            }
        });
        builder.setNegativeButton("CANCEL", null);

        AlertDialog dialog = builder.create();
        dialog.show();

        //show the keyboard when the dialog appears
        try {
            dialog.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
        } catch (NullPointerException e) {
            Log.e("failed to show keyboard", e.getMessage());
        }
    }
}

File: app/src/main/java/com/alderferstudios/noteappexample/MainActivity.kt
package com.alderferstudios.noteappexample

import android.content.Context
import android.content.DialogInterface
import android.content.SharedPreferences
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.ListView

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

import java.lang.reflect.Type
import java.util.ArrayList

class MainActivity : AppCompatActivity() {

    //the list of items
    private var listItems: MutableList<String>? = null
    private var adapter: ArrayAdapter<String>? = null

    private var prefs: SharedPreferences? = null
    private var editor: SharedPreferences.Editor? = null

    /**
     * Executes on app start
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setSupportActionBar(findViewById(R.id.toolbar) as Toolbar)

        prefs = getPreferences(Context.MODE_PRIVATE)
        editor = prefs!!.edit()

        //read in saved list
        val savedList = prefs!!.getString("list", "")
        if (savedList == "") { //nothing saved
            listItems = ArrayList<String>()
        } else {    //restore saved
            val type = object : TypeToken<ArrayList<String>>() {

            }.type
            listItems = Gson().fromJson<List<String>>(savedList, type) as MutableList<String>
        }

        val list = findViewById(R.id.list) as ListView

        // adapter to tie listItems to list
        adapter = ArrayAdapter(this,
                R.layout.list_row, R.id.listText, listItems!!)

        // assign the list adapter
        list.adapter = adapter
    }

    /**
     * Save list on app close
     */
    public override fun onStop() {
        super.onStop()

        //turn list into json
        val items = Gson().toJson(listItems)

        //save list to cache
        editor!!.putString("list", items)
        editor!!.apply() //use apply, its faster than commit
    }

    /**
     * Adds an item to the ListView
     * @param v the FAB
     */
    fun addItem(v: View) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Please enter the note to add")

        val input = EditText(this)
        builder.setView(input)

        builder.setPositiveButton("OK") { dialog, whichButton ->
            //add the input to the list
            listItems!!.add(input.text.toString())
            adapter!!.notifyDataSetChanged()
        }
        builder.setNegativeButton("CANCEL", null)

        val dialog = builder.create()
        dialog.show()

        //show the keyboard when the dialog appears
        try {
            dialog.window!!.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)
        } catch (e: NullPointerException) {
            Log.e("failed to show keyboard", e.message)
        }

    }
}
